_posts/2021-01-12-feedback-to-users-requests-in-distributed-systems.md

```diff
@@ -0,0 +1,105 @@
+
+Once the backplane is in place, the full communication flow is like the following:
+
+- The client:
+  - generates a unique identifier to use as correlation ID for a request
+  - creates the request payload and appends the identifier
+  - stores the identifier locally
+  - sends the request to the webserver
+- the webserver (instance)
+  - receives the user request 
+  - converts the user request into a message
+  - stores the received correlation ID into message headers alongside the connected WebSocket client ID
+  - sends the message
+  - sends back to the user an ACK (e.g., 202 Accepted)
+- the backend service (instance)
+  - receives the message
+  - processes the request
+  - creates the response message
+  - copies the correlation ID from the incoming message headers to the outgoing message headers
+  - send the response message
+- the webserver (instance):
+  - gets the correlation ID from the incoming message
+  - looks for the stored correlation ID and client WebSocket ID
+    - if it cannot find it, it means that someone else already dealt with the response, and this attempt is probably a duplicate message
+  - checks if the client is connected:
+    - if it is, uses the WebSocket connection to inform the client about the status
+    - if it's not, uses the backplane to propagate the received information to other web server instances.
+    
+### The elephant in the room
+
+One aspect that we haven't touched, it probably deserves its own set of articles, is how to handle results eventual consistency at the user interface level. My colleague Dennis Van Der Stelt does an excellent job in a [presentation he gave at Oredev](https://www.youtube.com/watch?v=jTz74m1KbBs).
+
+## Wrap up 
+
+As the reader can probably spot, a lot is going on. Handling a request asynchronously and, at the same time, providing feedback to the user brings a lot of complexity into the system. It's essential to evaluate if all this complexity is required; all the moving pieces need to be governed and maintained over time. One of the biggest lessons I've learned in my career is that using the same technique, architecture, or even technology in the entire system is the worst thing we can do. We'll end up with complexity where we don't need it and simplicity where too simple is harmful. Even if it might seem more complicated initially, it's worth evaluating each scenario and deciding how to design that specific scenario in isolation. I'm sure there will be a few scenarios where asynchronicity is required and many others where a much simpler approach is a better fit.
```

Comment:
```suggestion
As you've probably noticed by now, a lot is going on. Handling a request asynchronously and, at the same time, providing feedback to the user brings a lot of complexity into the system. It's essential to evaluate if all this complexity is required; all the moving pieces need to be governed and maintained over time. One of the biggest lessons I've learned in my career is that using the same technique, architecture, or even technology in the entire system is the worst thing we can do. We'll end up with complexity where we don't need it and simplicity where too simple is harmful. Even if it might seem more complicated initially, it's worth evaluating each scenario and deciding how to design that specific scenario in isolation. I'm sure there will be a few scenarios where asynchronicity is required and many others where a much simpler approach is a better fit.
```